### PR TITLE
In 1052 add guardrails final runs

### PR DESCRIPTION
**Pull request by jonavellecuerdo**
### Purpose and background context
Describe the overall purpose of the PR changes and any useful background context.

### How can a reviewer manually see the effects of these changes?
Explain how to see the proposed changes in the application if possible.

Delete this section if it isn't applicable to the PR.

### Includes new or updated dependencies?
YES | NO

### Changes expectations for external applications?
YES | NO

### What are the relevant tickets?
- Include links to Jira Software and/or Jira Service Management tickets here.

### Developer
- [ ] All new ENV is documented in README
- [ ] All new ENV has been added to staging and production environments
- [ ] All related Jira tickets are linked in commit message(s)
- [ ] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed and verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -50,16 +52,11 @@ def test_config_env_access_error(config):
...
-    monkeypatch.delenv("ALMA_SAP_INVOICES_ECS_NETWORK_CONFIG")
-    with pytest.raises(
-        InvalidNetworkConfigurationError,
-        match="the JSON object must be str, bytes or bytearray, not NoneType",
-    ):
-        _ = config.ALMA_SAP_INVOICES_ECS_NETWORK_CONFIG
+def test_config_env_access_network_config_success(monkeypatch, config):
+    assert config.ALMA_SAP_INVOICES_ECS_NETWORK_CONFIG == {
+        "awsvpcConfiguration": {
+            "subnets": ["subnet-abc123"],
```
Comment (cabutlermit) [inline tests/test_config.py:58]: @jonavellecuerdo it is more likely than not that the "subnets" value will be a list of subnets, not a single subnet. Would it make sense for your unit test to list two subnets instead of one here?

File: tests/test_config.py [hunk 2]
```diff
@@ -50,16 +52,11 @@ def test_config_env_access_error(config):
...
+            "securityGroups": ["sg-abc123"],
```
Comment (cabutlermit) [inline tests/test_config.py:59]: related to my comment on the preceding line, it is very unlikely that the "securityGroups" value will ever have anything more than just a single security group.

File: webapp/templates/process_invoices_confirm_final_run.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+{% block content %}
+<h1>Confirm final run</h1>
+<h3>Are you sure you want to execute a <em>final</em> run?</h2>
+
+<p>
+    <a class="btn button-primary" href="{{ url_for('process_invoices') }}">No</a>
+    <a class="btn button-primary warn" href="{{ url_for('process_invoices_run_execute', run_type='final') }}">
+        Execute a <strong>final</strong> run
+    </a>
```
Comment (ghukill) [inline webapp/templates/process_invoices_confirm_final_run.html:16]: I think this button is _right_ on the line of maybe too busy / too many emojis + colors.  But, I like it!  Perhaps stakeholders will have feedback about language, but it certainly adds a check to the process.

File: webapp/app.py [hunk 4]
```diff
@@ -123,16 +124,31 @@ def process_invoices() -> str:
...
+        elif run_type == "final":  # noqa: RET505
+            return redirect(url_for("process_invoices_confirm_final_run"))
+        return abort(400, description=f"Invalid run type: '{run_type}'")
+
+    @app.route("/process-invoices/run/final/confirm")
+    @login_required
+    def process_invoices_confirm_final_run() -> str:
+        return render_template("process_invoices_confirm_final_run.html")
+
+    @app.route("/process-invoices/run/<run_type>/execute")
```
Comment (ghukill) [inline webapp/app.py:139]: **Disclaimer:** Don't think changes are needed here at this time, but wanted to comment on this.

We have a route that provides a confirmation button in the UI, which is great, but that effectively just navigates you to this route (URL).  _This_ URL does the work of starting a "final" run, then redirects the user to the status page.

The good news: most casual users will probably never know this URL route exists given the immediate redirect of their browser.

The bad news: if people _do_ know this URL exists, and they intentionally / accidentally navigate straight to it, this kind of bypasses the confirmation page behavior.

At a high level, one option could be for the confirmation button to pass some kind of secret to this route, thereby informing it that it came from a confirmation (`GET` parameter if only URL, or could change to a `POST` route and have it submit like a form, or cookies, etc., lots of options).

But to restate: I _don't_ think this is needed now, and there is a balance with how gaurd-y these confirmations need to be.  It seems unlikely to me this pattern of someone hitting this route directly, but felt worth at least observing in this PR.

File: webapp/utils/aws/cloudwatch.py [hunk 5]
```diff
@@ -51,38 +51,17 @@ def get_log_events(self, task_id: str) -> list:
...
-            "logGroupName": self.log_group_name,
-        }
         while True:
-            response = self.client.describe_log_streams(**params)  # type: ignore[arg-type]
-            log_streams.extend(
-                logstream["logStreamName"] for logstream in response["logStreams"]
-            )
-            next_token = response.get("nextToken")
-            if next_token is None:
-                # all log streams retrieved is "nextToken" excluded from response
```
Comment (ghukill) [inline webapp/utils/aws/cloudwatch.py:85]: Cutting is delivering!  ❤️ 

In all seriousness, nice work here.  It's hard to cut code, especially new code that you just wrote.  But to my eyes, feels like you've found a more direct and straight-forward approach here.

File: webapp/app.py [hunk 6]
```diff
@@ -173,12 +173,13 @@ def process_invoices_status(task_id: str) -> str:
...
     @login_required
     def process_invoices_status_data(task_id: str) -> Response:
+        t_0 = time.time()
         try:
             task_status, logs = get_task_status_and_logs(task_id)
         except ECSTaskLogStreamDoesNotExistError:
             task_status = "UNKNOWN"
             logs = ["Log stream does not exist."]
-
+        logger.info(f"Data route elapsed: {time.time()-t_0}")
```
Comment (ghukill) [inline webapp/app.py:182]: 👍

File: webapp/config.py [hunk 7]
```diff
@@ -34,62 +32,29 @@ def __getattr__(self, name: str) -> Any:
...
-
-    @property
-    def ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP(self) -> str:
-        return self._getenv("ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP")
+        security_groups = self.ALMA_SAP_INVOICES_ECS_GROUPS
+        subnets = self.ALMA_SAP_INVOICES_ECS_SUBNETS
+        return {
+            "awsvpcConfiguration": {
+                "subnets": subnets.split(","),
+                "securityGroups": security_groups.split(","),
```
Comment (ghukill) [inline webapp/config.py:42]: What happens if someone includes a space between the values, e.g. `abc123, def456`?

Personally, I don't think it's worth dedicated properties that would split + strip the values... but felt worth mentioning.  As long as the README outlines the structure of these env vars I think we're okay!

File: README.md [hunk 8]
```diff
@@ -27,6 +27,8 @@ A Flask application run in AWS Lambda used to process invoices from Alma. Invoic
 ALMA_SAP_INVOICES_ECR_IMAGE_NAME=### The name of the ECR image for the Alma SAP Invoices CLI.
 ALMA_SAP_INVOICES_ECS_CLUSTER=### The full ARN of the ECS cluster to run the Alma SAP Invoices CLI ECS task.
 ALMA_SAP_INVOICES_ECS_TASK_DEFINITION=### The family and revision (formatted as <family>:<revision>) or full ARN of Alma SAP Invoices CLI ECS task. 
+ALMA_SAP_INVOICES_ECS_GROUPS=### Security group(s) for the Alma SAP Invoices ECS task (formatted as a comma-separated string).
+ALMA_SAP_INVOICES_ECS_SUBNETS=### Subnet(s) for the Alma SAP Invoices ECS task (formatted as a comma-separated string).
```
Comment (ghukill) [inline README.md:31]: Related to another comment: I think this suffices.  We use this comma-separated env var type in a few projects, and I think it's common knowledge that you should not include spaces.

File: webapp/utils/__init__.py [hunk 9]
```diff
@@ -52,14 +52,15 @@ def get_task_status_and_logs(task_id: str) -> tuple[str, list]:
...
-    logs = ["Waiting for logs."]
-    if task_status == "COMPLETED":
+    logs = ["Loading."]
+    if task_status in ["COMPLETED", "UNKNOWN"]:
         logs = cloudwatchlogs_client.get_log_messages(task_id)
+        if not logs:
+            return "EXPIRED (UNKNOWN)", ["Log stream expired, cannot find logs for task."]
 
-    return task_status, logs
+    return "COMPLETED", logs
```
Comment (ghukill) [inline webapp/utils/__init__.py:63]: Wondering if I'm missing something here?  How could the status be anything _but_ `COMPLETE`, `UNKNOWN`, or `EXPIRED (UNKNOWN)`?

It seems as though any other `task_status` will be overridden by this line that sets it to `COMPLETE`, as this is the only part of the method that returns a value.

Here is the full method from the PR branch:

```python
def get_task_status_and_logs(task_id: str) -> tuple[str, list]:
  # docstring here...
  ecs_client = ECSClient()
  cloudwatchlogs_client = CloudWatchLogsClient()
  # If task exists, get the current status
  try:
      task_status = (
          "COMPLETED"
          if (status := ecs_client.get_task_status(task_id)) == "STOPPED"
          else status
      )
  except ECSTaskDoesNotExistError:
      task_status = "UNKNOWN"
  
  logs = ["Loading."]
  if task_status in ["COMPLETED", "UNKNOWN"]:
      logs = cloudwatchlogs_client.get_log_messages(task_id)
      if not logs:
          return "EXPIRED (UNKNOWN)", ["Log stream expired, cannot find logs for task."]
  
  return "COMPLETED", logs
```

I ran a "review" run in the deployed version, and ended up with this state, which doesn't feel correct:

Screenshot 2024-09-04 at 9 36 34 AM

**Review by ghukill — CHANGES_REQUESTED** [review]
Overall, looking good.  I particularly appreciate the detailed PR notes and examples.

Left a couple questions that likely could be resolved later, or not at all.

I did leave [one question](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83#discussion_r1743817463) which prompted this "Request Changes" review, as it _seems_ like it could be a bug?  Unless not understanding it?

File: webapp/utils/aws/cloudwatch.py (hunk 5 above)
Comment (ehanson8) [reply webapp/utils/aws/cloudwatch.py:85]: Yes, this feels more streamlined!

**Review by ehanson8 — COMMENTED** [review]
Good changes and the context in the PR description is greatly appreciated as always! I'm also curious about the issues @ghukill raised so just commenting for now

File: webapp/config.py (hunk 7 above)
Comment (jonavellecuerdo) [reply webapp/config.py:42]: `boto3` returns an `InvalidParameterException` for incorrectly formatted parameter values. I agree that formatting errors can be mitigated by a detailed README description for the env var. The descriptions will be updated to: 

```shell
ALMA_SAP_INVOICES_ECS_GROUPS=### Security group(s) for the Alma SAP Invoices ECS task (formatted as a comma-separated string excluding whitespaces, e.g. "sg-abc123,sg-def456").
ALMA_SAP_INVOICES_ECS_SUBNETS=### Subnet(s) for the Alma SAP Invoices ECS task (formatted as a comma-separated string excluding whitespaces, e.g. "subnet-abc123,subnet-def456").
```

File: README.md (hunk 8 above)
Comment (jonavellecuerdo) [reply README.md:31]: See response in https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83#discussion_r1744067445.

File: tests/test_config.py (hunk 1 above)
Comment (jonavellecuerdo) [reply tests/test_config.py]: Thank you for the suggestion! See this commit https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83/commits/c858fd122d3337e7d3016e81940011f6e50229e0 for more details. 

**Note:** [`boto3.ECSClient.run_task`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs/client/run_task.html) requires these network configuration dictionary to pass these values as a list. The call to `split` the string by the comma (`,`) will return a list, even for single values, so I think [this logic](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83/files#diff-966b49fcc05ab61f6b2ef9d13713ceb9da0b5d825b9ef5cdb16961b931e6bfd6R41-R42) is fine to leave as-is.

File: tests/test_config.py (hunk 2 above)
Comment (jonavellecuerdo) [reply tests/test_config.py:59]: See response in https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83#discussion_r1742667164. 😄

File: webapp/config.py (hunk 7 above)
Comment (jonavellecuerdo) [reply webapp/config.py:42]: See [updates in README](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R30-R31).

File: webapp/templates/process_invoices_confirm_final_run.html (hunk 3 above)
Comment (ghukill) [reply webapp/templates/process_invoices_confirm_final_run.html:16]: @jonavellecuerdo - good point red color blindness!  I had not considered that.  +100 to style guide usage + thoughtful implementation.  I hereby retract all "busy" concerns.

File: webapp/utils/__init__.py (hunk 9 above)
Comment (jonavellecuerdo) [reply webapp/utils/__init__.py]: Ah, I don't have a good reason for this error other than it being a misstep in an attempt to fix a recurring call to the "data" route. 😅 One of the updates I made resulted in a never-endling loop that sent a request to the "data" route -- even after the logs had been retrieved. I've fixed the bug via https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83/commits/8e3743744c9438395f0001314bc199dbfc233bf3 and re-deployed to Dev1.

File: webapp/app.py (hunk 4 above)
Comment (jonavellecuerdo) [reply webapp/app.py:139]: Thank you for pointing this out! I'll start looking into how we can fortify the confirmation workflow. 🤔

Comment (jonavellecuerdo) [reply webapp/app.py:140]: @ghukill @ehanson8 Please see the proposed solution in https://github.com/MITLibraries/alma-sapinvoices-ui/pull/83/commits/1ca21d1f6a39ae182d28030c7f6e6df38145cf2c. The 'execute' route for final runs (i.e., `/process-invoices/run/final/execute` is now protected against direct get requests by using the [Flask.request.referrer](https://flask.palletsprojects.com/en/2.3.x/api/#flask.Request.referrer) to confirm that the user was redirected by the confirmation page.

Comment (ghukill) [reply webapp/app.py:140]: @jonavellecuerdo - awesome idea.  As mentioned in a huddle, to me this strikes a nice balance.  It provides a bit more gaurdrail in the sense you _must_ come from that button confirmation page for this to happen... but it's not quite at the level of `POST` requests + forms + secret, hidden keys getting passed around, which doesn't feel necessary.  Nice work!

File: webapp/templates/process_invoices_confirm_final_run.html (hunk 3 above)
Comment (jonavellecuerdo) [reply webapp/templates/process_invoices_confirm_final_run.html:16]: The style guide has been helpful throughout this process! :) The "busy" feeling could be that the phrase on the button is a bit long, but I think a little wordiness is fine considering we're _really_ trying to prevent "final" runs. 😅

**Review by ehanson8 — APPROVED** [review]
Looks good!

**Review by ghukill — APPROVED** [review]
Looking great to me.

I left a comment around the `get_task_status_and_logs` method, as I did find it a little hard to follow, but I'm okay approving without changes there; it's mostly just stylistic / readability, and could be a personal preference.

File: webapp/utils/__init__.py [hunk 10]
```diff
@@ -34,13 +34,21 @@ def get_task_status_and_logs(task_id: str) -> tuple[str, list]:
...
+          - If the log stream exists and is not expired, the method will return
+            a tuple with task_status = "COMPLETED" and the logs.
+          - If the log stream exists but is expired, the method will return a
+            tuple with task_status = "EXPIRED (UNKNOWN)" and a corresponding message.
+       - If task_status = "COMPLETED", retrieve logs. This ensures the full set of logs
+         are retrieved only when the task run completed.
+       - If the log stream does not exist, CloudWatchLogsClient.get_log_messages
+         raises ECSTaskLogStreamDoesNotExistError.
     """
     ecs_client = ECSClient()
```
[the remark this answers is not in the document]
Comment (jonavellecuerdo) [reply webapp/utils/__init__.py:53]: Hmm, I think I'll leave it as is for now. Although the current syntax might be a bit longer, I like that it separates the logic for "old" ECS tasks vs. "recent" ECS tasks!

[the remark this answers is not in the document]
Comment (ghukill) [reply webapp/utils/__init__.py:53]: Rightyo!  Makes sense.  Thanks for considering.
